### ISSUE TITLE
QT5 : use of deleted function

### ISSUE THREAD

**Issue by vincentvije** [veyon/veyon#36]
Hello,

I compiled in a fresh installed Tumbleweed.
I get:
[ 21%] Building CXX object core/CMakeFiles/veyon-core.dir/src/DesktopAccessDialog.cpp.o
In file included from /usr/include/qt5/QtCore/qobject.h:49:0,
                 from /usr/include/qt5/QtCore/qcoreapplication.h:46,
                 from /usr/include/qt5/QtCore/QCoreApplication:1,
                 from /root/Downloads/veyon-3.99.6/core/src/DesktopAccessDialog.cpp:25:
/usr/include/qt5/QtCore/qlist.h: In instantiation of 'void QList<T>::node_construct(QList<T>::Node*, const T&) [with T = Feature]':
/usr/include/qt5/QtCore/qlist.h:584:27:   required from 'void QList<T>::append(const T&) [with T = Feature]'
/usr/include/qt5/QtCore/qlist.h:384:13:   required from 'QList<T>& QList<T>::operator+=(const T&) [with T = Feature]'
/root/Downloads/veyon-3.99.6/core/src/DesktopAccessDialog.cpp:43:16:   required from here
/usr/include/qt5/QtCore/qlist.h:441:35: error: use of deleted function 'Feature& Feature::operator=(const Feature&)'
     else *reinterpret_cast<T*>(n) = t;
          ~~~~~~~~~~~~~~~~~~~~~~~~~^~~
In file included from /root/Downloads/veyon-3.99.6/core/include/FeatureMessage.h:30:0,
                 from /root/Downloads/veyon-3.99.6/core/include/FeaturePluginInterface.h:29,
                 from /root/Downloads/veyon-3.99.6/core/include/DesktopAccessDialog.h:30,
                 from /root/Downloads/veyon-3.99.6/core/src/DesktopAccessDialog.cpp:30:
/root/Downloads/veyon-3.99.6/core/include/Feature.h:35:25: note: 'Feature& Feature::operator=(const Feature&)' is implicitly deleted because the default definition would be ill-formed:
 class VEYON_CORE_EXPORT Feature : public QObject
                         ^~~~~~~
/root/Downloads/veyon-3.99.6/core/include/Feature.h:35:25: error: 'QObject& QObject::operator=(const QObject&)' is private within this context
In file included from /usr/include/qt5/QtCore/qcoreapplication.h:43:0,
                 from /usr/include/qt5/QtCore/QCoreApplication:1,
                 from /root/Downloads/veyon-3.99.6/core/src/DesktopAccessDialog.cpp:25:
/usr/include/qt5/QtCore/qobject.h:454:5: note: declared private here
     Q_DISABLE_COPY(QObject)
     ^
In file included from /root/Downloads/veyon-3.99.6/core/include/FeatureMessage.h:30:0,
                 from /root/Downloads/veyon-3.99.6/core/include/FeaturePluginInterface.h:29,
                 from /root/Downloads/veyon-3.99.6/core/include/DesktopAccessDialog.h:30,
                 from /root/Downloads/veyon-3.99.6/core/src/DesktopAccessDialog.cpp:30:
/root/Downloads/veyon-3.99.6/core/include/Feature.h:35:25: error: use of deleted function 'QObject& QObject::operator=(const QObject&)'
 class VEYON_CORE_EXPORT Feature : public QObject
                         ^~~~~~~
In file included from /usr/include/qt5/QtCore/qcoreapplication.h:43:0,
                 from /usr/include/qt5/QtCore/QCoreApplication:1,
                 from /root/Downloads/veyon-3.99.6/core/src/DesktopAccessDialog.cpp:25:
/usr/include/qt5/QtCore/qobject.h:454:5: note: declared here
     Q_DISABLE_COPY(QObject)
     ^
make[2]: *** [core/CMakeFiles/veyon-core.dir/build.make:654: core/CMakeFiles/veyon-core.dir/src/DesktopAccessDialog.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:2278: core/CMakeFiles/veyon-core.dir/all] Error 2
make: *** [Makefile:152: all] Error 2

Thanks for your work and your help.
Kind regards,

Vincent

**Comment by tobydox** [comment]
Thanks for the feedback! I assume you're using Qt 5.9?

**Comment by vincentvije** [comment]
Hello Tobias,
Yes. Qt5Core 5.9.0-1.1
If you need me to test or make something, just ask me.
Thank you very much,
Vincent

**Comment by vincentvije** [comment]
Hello again,

I've computers under OpenSuSE 13.1, and italc 2 works fine, and lock and 
unlock in less than 1 second !

But under these same computers upgraded to Leap 42.1 and 42.2, when I 
ask the master to lock a screen, it takes 10 seconds. And to unblock 10 
seconds again.
I upgraded to italc 3.0.3, but I get the same problem.

Do you have any idea ?

Under Tumbleweed freshly installed computer, I've the problem you 
already saw with Veyon install.
But with talc 3.0.3, lock and unlock don't work anymore, it just freeze 
the computer...

I use OpenSuSE since 20 years, but now I looking to use another 
distribution...

Thanks for any help or suggestion about Veyon, or about these 10 seconds 
that takes italc to lovk and unlock.

Vincent

**Comment by tobydox** [comment]
Yeah the screen lock functionality (and similar functions using IPC mechanisms) is fundamentally broken in iTALC 3, probably due to Qt 5. However in Veyon most of the core has been rewritten so Veyon doesn't suffer from these problems any longer. We should therefore concentrate on getting Veyon running in your environment :-)

**Comment by vincentvije** [comment]
Thanks for the explanation, Tobias.
Yes, please if you need me make some tests, new compilation, and so on, 
please just ask me.
I'm not developper, but ready to help where I can.
Kind regards,
Vincent MARECHAL

**Comment by tobydox** [comment]
I just committed a fix, can you please pull the latest sources and try again?

**Comment by vincentvije** [comment]
*Hello Tobias,*

*Ok got another X11vnc source file, and created a .rpm. Great !**
*I installed it, I got some problems with dependencies, but I think it's 
fine now.

_*Installing an configuring both sides :*__*
*_

    I installed Veyon for Windows 64.
    I configured certificates in both sides.
    In Veyon for Windows, Master side, in the config window, I added my
    Linux Tumbleweed computer.
    Then I started Veyon Master.

*_Veyon Master :_*

    I can see in the Veyon Master the live thumnail of my Tumbleweed - Fine
    I can select it and make a Remote Control - Fine

    *But when I select the Tumbleweed computer and click "Lock", then
    like in italc, it freeze the tumbleweed computer : I can see the
    freezed screen as before cliking "Lock".**
    *If then I click "Unlock", it never come back to life ! But after a
    lot of seconds, I can make "ctrl" + "alt" + "F1", and open a
    terminal session to reboot my tumbleweed computer...

    So it react the same as italc : "Lock" freeze the X11 of the client,
    "Unlock" allow after dozens of seconds to get back keyboard and
    mouse pointer but the mouse button do not interact anymore..., wih
    the keyboard, I go to terminal ("ctrl" + "alt" + "F1") and just ask
    the computer to reboot.

For the moment, I'm stuck here.

I hope it can help you.
I've no idea of what to do.

Thanks again.
Kind regards,

Vincent

**Comment by vincentvije** [comment]
I add : see this screenshot : I make "Lock" then immediately "Unlock".
I get the control back after 6 min.
The network connection is very fast on both ends...

**Comment by tobydox** [comment]
Hm maybe the call to xmodmap (which implements the keyboard locking) is the culprit here. Can you set log level to debug, clear log files on the tumbleweed computer (and restart service) and grab all log files after activating and deactivating the "Lock" function? You can then either attach the log files here or send me a PM.

**Comment by vincentvije** [comment]
Hello Tobias,

Well, I worked some more hours in trying to make it work, or to clone my 
install on a different computer, but get always the same problems.
Do you think that after 20 years of opensuse I should switch to ubuntu ?

Thanks a lot.
Kind regards,

Vincent

**Comment by tobydox** [comment]
At least you could give it a try. I'm really wondering why this doesn't work in your setup. I'll grab an opensuse tumbleweed VM and test with it.

**Comment by vincentvije** [comment]
Hello Tobias,

I will do it asap.

But when I "Lock", _I never have the padlock picture_ on the screen, 
just a frozen screen.
When I "Unlock", _I keep the frozen screen some minutes_, and when it 
goes back, the Keyboard stay disabled.

So, are you sure the problem is from keyboard ?
Could the kayboard keep computer frozen some minutes, and prevent the 
padlock to show ?

Kind regards,

Vincent

**Comment by vincentvije** [comment]
*Hello Tobias,*

_*Yes !*_

    I installed a fresh Tumbleweed, and all packages that I think valuable.
    I made all updates.

    Then I compiled and installed Veyon.

    On my windows 10, I installed Veyon, for the master side.

    Now it's working...

_*But...*_

    Nearly all the same than with italc, with "Lock" and "Unlock" (since
    Leap or Tumbleweed).
    They takes about 11 seconds to make effect : when I click "Lock" or
    "Unlock", it takes 11 seconds to get or remove the padlock on the
    client computer.

    The good point is that when I click "Lock", the student can't do
    anything immediately, but he wait 14 second to see the padlock.
    Other good point : the keyboard is inactive during the padlocked
    screen ! (with italc, the keyboard was functioning without being
    able to see the effects).
    And the same when I click "Unlock", the student wait 14 second to
    get the the padlock removed from the screen.

    Note that if the screen has been locked by the user, or the timer,
    Lock and Unlock work (with 11 seconds latency), but the padlock
    never shows.

    Note too that my installation of Veyon on the client side freeze
    like before (see previous email) on a virtual machine (vmware) with
    Tumbleweed, but works if I clone the vm on a physical computer with
    clonezilla for example. That's what I did too.

I hope this will help !
Do not hesitate if I can do something more.

For the moment, I'm stuck with this total latency of 22 seconds.
The problem is more when unlocking screen...

But we can say that it works well like itacl with Tumbleweed (but with 
the same latency too).

Thanks a lot for this great job you already done.
I hope we can solve this latency.

I'm in the idea to test it with a debian 9 installation. Did you already 
done this ?

Kind regards,

Vincent

**Comment by vincentvije** [comment]
*Hello Tobias,*

I just saw you release the version 4 of Veyon.
And you wrote you corrected the problems with Qt5.5 and the lag when 
locking/unlocking screen.

I just took your new rpm, and installed it on my opensuse 42.2, and 
installed the master on my windows.
I just followed then the same steps as in june, when I compiled in on 
linux, and got it working with lag between windows master and linux client.

But now I can't get it working...
I always get this error :


I actived only the auth with certificates, and replicated the 
certificates in both client and master, with the good paths.

If you have some advice, it could help me.
If not, then I will try to compile instead of using the rpm and will see 
if it helps me.

Thanks for this great job !
Kind regards,

Vincent

**Comment by vincentvije** [comment]
Hello Tobias,

I spent three more hour to try to make work Veyon 4.0.0 :
Master : Windows 10
Client : Opensuse Leap 42.2 with your rpm

I tried all possibilities, really. No way.
I tworked in june. Now it will not work.

I tried my certificates that worked in june.
I tried new certificates.
I tried auth with logon.
I installed a new fresh Leap 42.2.

Here is the "loop error" a have with old or new certificates, or logon, 
when I start veyon-service on the Leap client and then veyon master on 
windows 10 :


I hope you can help.
Thanks in advance,

Vincent
